File: AISandboxDev/03_Frontend/src/types/review.ts
export type ReviewDecision =
  | 'approved'
  | 'approved_with_controls'
  | 'restricted'
  | 'reassessment_required'
  | 'not_approved';

export type ReviewStatus = 'pending' | 'decided' | 'overridden';

export interface Review {
  id: string;
  modelId: string;
  modelName: string;
  modelProvider: string;
  runId: string;
  runDate: string;
  overallScore: number;
  criticalCount: number;
  status: ReviewStatus;
  decision?: ReviewDecision;
  decisionReason?: string;
  reviewerNotes?: string;
  reviewerId?: string;
  reviewerName?: string;
  isPublished: boolean;
  createdAt: string;
  updatedAt: string;
  auditTrail: AuditTrailEntry[];
}

export interface AuditTrailEntry {
  id: string;
  action: 'decision_set' | 'decision_overridden' | 'publication_toggled' | 'notes_updated';
  actor: string;
  actorName: string;
  timestamp: string;
  oldValue?: string;
  newValue?: string;
  reason?: string;
}

File: AISandboxDev/06_Backend/app/models/review.py
"""Review and AuditTrailEntry ORM models."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone

from sqlalchemy import String, Text, Float, Integer, Boolean, DateTime, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Review(Base):
    """Review record for a completed benchmark run."""

    __tablename__ = "reviews"

    id: Mapped[str] = mapped_column(String(50), primary_key=True, default=lambda: str(uuid.uuid4()))
    model_id: Mapped[str] = mapped_column(String(50), ForeignKey("sandbox_models.id"), nullable=False)
    model_name: Mapped[str] = mapped_column(String(255), nullable=False)
    model_provider: Mapped[str] = mapped_column(String(255), nullable=False)
    run_id: Mapped[str] = mapped_column(String(50), ForeignKey("runs.id"), nullable=False)
    run_date: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    overall_score: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    critical_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="pending")
    decision: Mapped[str | None] = mapped_column(String(50), nullable=True)
    decision_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    reviewer_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    reviewer_id: Mapped[str | None] = mapped_column(String(50), ForeignKey("users.id"), nullable=True)
    reviewer_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    is_published: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )


class AuditTrailEntry(Base):
    """Immutable audit log entry for review decisions."""

    __tablename__ = "audit_trail"

    id: Mapped[str] = mapped_column(String(50), primary_key=True, default=lambda: str(uuid.uuid4()))
    review_id: Mapped[str] = mapped_column(String(50), ForeignKey("reviews.id"), nullable=False)
    action: Mapped[str] = mapped_column(String(50), nullable=False)
    actor: Mapped[str] = mapped_column(String(50), ForeignKey("users.id"), nullable=False)
    actor_name: Mapped[str] = mapped_column(String(255), nullable=False)
    timestamp: Mapped[datetime] = mapped_column(
        DateTime, default=lambda: datetime.now(timezone.utc)
    )
    old_value: Mapped[str | None] = mapped_column(Text, nullable=True)
    new_value: Mapped[str | None] = mapped_column(Text, nullable=True)
    reason: Mapped[str | None] = mapped_column(Text, nullable=True)
